File: EngineKitGame/EngineKit/EngineKit/Items/Shape/Pyramid.m


#import "Pyramid.h"


@implementation Pyramid

+ (instancetype)pyramid {
    return [self create];
}

- (instancetype)init {
    if (self = [super init]) {
        self.pyramid = [SCNPyramid new];
    }
    return self;
}

- (instancetype)initWithWidth:(CGFloat)width
                       height:(CGFloat)height
                       length:(CGFloat)length {
    if (self = [super init]) {
        self.pyramid =
            [SCNPyramid pyramidWithWidth:width height:height length:length];
    }
    return self;
}

- (Item *)deepCopy {
    Shape *newItem = [[self class] new];
    [self copyInfoTo:newItem];
    [super copyPhysicsTo:newItem];
    return newItem;
}

- (void)copyInfoTo:(Pyramid *)item {
    [super copyInfoTo:item];
    
    item.width = self.width;
    item.height = self.height;
    item.length = self.length;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Property Overriding

- (void)setPyramid:(SCNPyramid *)pyramid {
    self.geometry = pyramid;
}

- (SCNPyramid *)pyramid {
    return (SCNPyramid *)self.geometry;
}


- (void)setWidth:(NSNumber *)width {
    [self assertTheresNoPhysicsBody];
    self.pyramid.width = width.doubleValue;
}

- (NSNumber *)width {
    return @(self.pyramid.width);
}

- (void)setHeight:(NSNumber *)height {
    [self assertTheresNoPhysicsBody];
    self.pyramid.height = height.doubleValue;
}

- (NSNumber *)height {
    return @(self.pyramid.height);
}

- (void)setLength:(NSNumber *)length {
    [self assertTheresNoPhysicsBody];
    self.pyramid.length = length.doubleValue;
}

- (NSNumber *)length {
    return @(self.pyramid.length);
}

@end

File: Project/Project/Items/Shapes/Pyramid.swift

import Foundation
import SceneKit
import JavaScriptCore




@objc class Pyramid: Shape {
    
    
    var width: CGFloat {
        set { pyramid.width = newValue }
        get { return pyramid.width     }
    }
    
    var height: CGFloat {
        set { pyramid.height = newValue }
        get { return pyramid.height     }
    }
    
    var length: CGFloat {
        set { pyramid.length = newValue }
        get { return pyramid.length     }
    }
    
    override var size: CGFloat {
        set {
            width = newValue
            height = newValue
            length = newValue
        }
        get {
            return max(max (width, height), length)
        }
    }
    
    var pyramid: SCNPyramid {
        set { geometry = newValue        }
        get { return geometry as SCNPyramid }
    }
    
    override init() {
        super.init(geometry: SCNPyramid())
    }
}

